elastic/beats
# Feature/remove k8s cache

## What does this PR do?

Replaced time expiring cache in the Kubernetes module with a dictionary that never expires. 

Since some metrics in this internal cache were only set once at startup and not regularly updated, we needed to replace the internal cache with a dictionary that never expires.

## Why is it important?

It fixes an issue that caused some metrics `kubernetes.container.cpu.usage.limit.pct` and `kubernetes.container.memory.usage.limit.pct`, `kubernetes.pod.memory.usage.limit.pct` and `kubernetes.pod.cpu.usage.limit.pct` to be missing at times:
- if state_node, node metricsets are disabled, 
  - container with limits set
    - `kubernetes.container.*` metrics are present
  - container without limits set
    - `kubernetes.container.*` metrics are missing. This is CORRECT since limits are not set on containers and we cannot use node metrics as backup option
  - pod with all container limits set
    - `kubernetes.pod.*` metrics are missing (Bug to be fixed here)
  - pod with all containers without limits set
    - `kubernetes.pod.*` metrics are missing. This is CORRECT since limits are not set on containers and we cannot use node metrics as backup option
- if state_node, node metricsets are enabled,
  - container/pod with limits set
    - all 4 metrics are present
  - container/pod without limits set
    - all 4 metrics are present but they use node limits instead of container limits.  This is CORRECT since limits are not set on containers and we use node metrics as backup option

## Checklist


## Author's Checklist

- if state_node, node metricsets are disabled, 
  - container/pod with limit sets
    - all 4 metrics are present
  - container/pod without limit sets
    -  all 4 metrics  metrics are missing. This is CORRECT since limits are not set on containers and we cannot use node metrics as backup option
- if state_node, node metricsets are enabled,
  - container/pod with limit sets
    -  all 4 metrics metrics are present
  - container/pod without limit sets
    -  all 4 metrics metrics are present but they use node limits instead of container limits. This is CORRECT since limits are not set on containers and we use node metrics as backup option

## How to test this PR locally

- Run metricbeat on kubernetes with either Kind with multiple nodes or a k8s cluster on GCP/AWS
- Add/remove test pods and see metrics appear/disappear for those pods using this manifest 
[nginx-daemons.yaml.zip](https://github.com/elastic/beats/files/9298250/nginx-daemons.yaml.zip)
- import the following [dashboard.ndjson.zip](https://github.com/elastic/beats/files/9292774/dashboard.ndjson.zip) and check for metrics according to `author's checklist`




## Related issues

- Closes https://github.com/elastic/beats/issues/32232

## Use cases



## Screenshots

When a pod is deleted (like in the following screenshot at 11:09:37) you might see the container still reporting a cpu/memory usage but not the limit.pct. When the pod is added back (like in the following screenshot at 11:12:31) you might see the container with cpu usage = 0 and limit.pct = 0 for a few seconds until the cpu is reported again and limit.pct to be different from 0.

![Screenshot 2022-08-01 at 11 13 21](https://user-images.githubusercontent.com/2378291/182126929-71b009dc-bc06-4014-b223-9a2006587946.png)


## Logs
